File: kora-app-annotation-processor/src/main/java/ru/tinkoff/kora/kora/app/annotation/processor/component/ResolvedComponents.java
package ru.tinkoff.kora.kora.app.annotation.processor.component;

import org.jspecify.annotations.Nullable;
import ru.tinkoff.kora.kora.app.annotation.processor.declaration.ComponentDeclaration;
import ru.tinkoff.kora.kora.app.annotation.processor.declaration.DeclarationWithIndex;

import java.util.Collection;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public class ResolvedComponents {
    private final Map<Integer, ResolvedComponent> declarationIndexToResolvedComponent = new LinkedHashMap<>();

    public ResolvedComponents() {
    }

    public ResolvedComponents(ResolvedComponents that) {
        this.declarationIndexToResolvedComponent.putAll(that.declarationIndexToResolvedComponent);
    }

    @Nullable
    public ResolvedComponent getByDeclarationIndex(int idx) {
        return declarationIndexToResolvedComponent.get(idx);
    }

    @Nullable
    public ResolvedComponent getByDeclaration(DeclarationWithIndex d) {
        return declarationIndexToResolvedComponent.get(d.index());
    }

    public void add(int declarationIdx, ComponentDeclaration declaration, List<ComponentDependency> resolvedDependencies) {
        declarationIndexToResolvedComponent.put(declarationIdx, new ResolvedComponent(
            declarationIndexToResolvedComponent.size(), declaration, declaration.type(), declaration.tag(),
            List.of(), // TODO,
            resolvedDependencies
        ));
    }

    public Collection<ResolvedComponent> components() {
        return declarationIndexToResolvedComponent.values();
    }

    public int size() {
        return declarationIndexToResolvedComponent.size();
    }
}

File: kora-app-symbol-processor/src/main/kotlin/ru/tinkoff/kora/kora/app/ksp/component/ResolvedComponents.kt
package ru.tinkoff.kora.kora.app.ksp.component

import ru.tinkoff.kora.kora.app.ksp.declaration.ComponentDeclaration
import ru.tinkoff.kora.kora.app.ksp.declaration.DeclarationWithIndex


class ResolvedComponents() {
    private val declarationIndexToResolvedComponent = LinkedHashMap<Int, ResolvedComponent>()
    val size get() = declarationIndexToResolvedComponent.size

    constructor(that: ResolvedComponents) : this() {
        declarationIndexToResolvedComponent.putAll(that.declarationIndexToResolvedComponent)
    }


    fun getByDeclarationIndex(idx: Int): ResolvedComponent? {
        return declarationIndexToResolvedComponent[idx]
    }

    fun getByDeclaration(d: DeclarationWithIndex): ResolvedComponent? {
        return declarationIndexToResolvedComponent[d.index]
    }

    fun add(declarationIdx: Int, declaration: ComponentDeclaration, resolvedDependencies: List<ComponentDependency>) {
        declarationIndexToResolvedComponent[declarationIdx] = ResolvedComponent(
            declarationIndexToResolvedComponent.size, declaration, declaration.type, declaration.tag, resolvedDependencies
        )
    }

    fun components(): Collection<ResolvedComponent> {
        return declarationIndexToResolvedComponent.values
    }
}
